training/src/anemoi/training/checkpoint/catalog.py

```diff
@@ -0,0 +1,657 @@
+# (C) Copyright 2024 Anemoi contributors.
+#
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# In applying this licence, ECMWF does not waive the privileges and immunities
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+
+"""Component catalog for pipeline components with automatic discovery.
+
+This module provides automatic discovery of checkpoint pipeline components
+through module inspection. It dynamically finds all available sources,
+loaders, and modifiers by scanning the package structure.
+
+The catalog discovers components by:
+- Scanning the appropriate modules (sources, loaders, modifiers)
+- Finding all classes that inherit from the base classes
+- Building a registry automatically at import time
+
+This provides true discovery without hardcoding, making it easy to add
+new components without updating the catalog.
+
+Example
+-------
+>>> from anemoi.training.checkpoint.catalog import ComponentCatalog
+>>>
+>>> # List dynamically discovered components
+>>> print(ComponentCatalog.list_sources())
+>>> ['local', 's3', 'http', 'gcs', 'azure']
+>>>
+>>> # Get component target path
+>>> target = ComponentCatalog.get_source_target('s3')
+>>> print(target)
+>>> 'anemoi.training.checkpoint.sources.S3Source'
+>>>
+>>> # Create component using standard Hydra config
+>>> from hydra.utils import instantiate
+>>> config = {
+...     '_target_': 'anemoi.training.checkpoint.sources.S3Source',
+...     'bucket': 'my-bucket',
+...     'key': 'model.ckpt'
+... }
+>>> source = instantiate(config)
+"""
+
+from __future__ import annotations
+
+import importlib
+import inspect
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class ComponentCatalog:
```

Comment:
Seems like useful functionality but I wonder if this should live in anemoi-utils or potentially https://github.com/ecmwf/anemoi-registry/tree/main?

(as in how much of this is for classes related to checkpoints or extensive to any other anemoi class?)

Reply:
Also it could be interesting to get @gareth-j and @jjlk takes as if this could be relevant in the context or the anemoi-catalogue?